### PR TITLE
Introduce automaton request throttler for ObjectService.Execute write subrequests

### PR DESCRIPTION
**Pull request by achulkov2**
This PR introduces a new throttler in ObjectService that allows limiting the total rate of write subrequests from all non-root users. The general intent is to use this throttler in combination with existing request-fairness by user to refrain from using per-user throttlers.

It also adds some tests for existing read/write user throttlers, which should be extended in the future to cover other scenarios (e.g. distributed throttler, single master, multicell).

### CODE REVIEW

File: yt/yt/server/master/object_server/object_service.cpp [hunk 1]
```diff
@@ -1596,9 +1599,20 @@ class TObjectService::TExecuteSession
...
-                if (!WaitForAndContinue(result)) {
+                TFuture<void> automatonThrottlerFuture = VoidFuture;
+                if (subrequestType == EExecutionSessionSubrequestType::LocalWrite && User_.Get() != securityManager->GetRootUser()) {
+                    automatonThrottlerFuture = Owner_->AutomatonRequestThrottler_->Throttle(1);
+                }
+
+                // NB: This callback may be executed in an arbitrary thread.
+                auto throttlingResult = automatonThrottlerFuture.Apply(
+                    BIND_NO_PROPAGATE([=, this, this_ = MakeStrong(this), user = User_.Get()] () {
+                        const auto& securityManager = Bootstrap_->GetSecurityManager();
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1611]: I found the ownership semantics in this case to be too cumbersome. There is an underlying promise in `AllSet`, which we would need to keep alive past `WaitForAndContinue`.

**Review by gritukan — APPROVED** [review]

File: yt/yt/tests/integration/master/test_users.py [hunk 2]
```diff
@@ -818,6 +822,148 @@ def test_password_temporary(self):
...
+
+        request_results = execute_batch([params["make_batch_request_function"](i) for i in range(params["request_count"])], authenticated_user=user)
+        for request_result in request_results:
+            raise_batch_error(request_result)
+
+        no_throttling_batch_execution_time = time.time() - start_time
+
+        throttled_batch_execution_time = 0
+
+        with deferred(params["throttling_limit_reset_function"]):
```
Comment (raundn) [inline yt/yt/tests/integration/master/test_users.py:914]: Shouldn't teardown code handle this? It definitely resets the dynconfig. I'm not so sure about per-user throttlers but I suspect all the users should be removed at teardown anyway.

**Review by raundn — CHANGES_REQUESTED** [review]

File: yt/yt/server/master/security_server/request_tracker.cpp [hunk 3]
```diff
@@ -172,7 +172,8 @@ void TRequestTracker::ReconfigureUserRequestRateThrottlers(TUser* user)
         config->Period = GetDynamicConfig()->RequestRateSmoothingPeriod;
 
         auto cellTag = Bootstrap_->GetMulticellManager()->GetCellTag();
-        auto requestRateLimit = user->GetRequestRateLimit(workloadType, cellTag);
+        // Conversion to double is needed due to potential division below.
+        std::optional<double> requestRateLimit = user->GetRequestRateLimit(workloadType, cellTag);
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/security_server/request_tracker.cpp:176]: Well, `read_request_rate_limit = 1` never worked properly: 1 / (2+ followers) turned into 0 in both throttlers and (even worse) there was a [bug](https://github.com/ytsaurus/ytsaurus/pull/700), which caused limit = 0 to be interpreted as effectively infinite. 
I could have used limit = 2 in tests, but I figured it is best to fix the bugs instead.

File: yt/yt/server/master/object_server/object_service.cpp [hunk 4]
```diff
@@ -1600,9 +1603,20 @@ class TObjectService::TExecuteSession
...
                 const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
 
-                if (!WaitForAndContinue(result)) {
+                auto automatonThrottlerFuture = VoidFuture;
+                if (subrequestType == EExecutionSessionSubrequestType::LocalWrite && User_.Get() != securityManager->GetRootUser()) {
+                    automatonThrottlerFuture = Owner_->AutomatonRequestThrottler_->Throttle(1);
+                }
+
+                // NB: This callback may be executed in an arbitrary thread.
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1612]: It is definitely fragile, but not sure how avoidable that is.

Re arbitrary thread: we either rely on `securityManager->ThrottleUser` being thread-safe (which, in the end, it [is](https://github.com/ytsaurus/ytsaurus/blob/main/yt/yt/server/master/security_server/user.cpp#L462)), or we drag this callback through the current invoker (besides `Automaton` this can also be a `LocalRead` thread), which isn't that efficient. 

User pointer should not be dead because we capture a strong ref to `this`, which contains `TEphemeralObjectPtr<TUser> User_`. I asked Grisha about it, and he told me that it should guarantee that the user is alive. Maybe I misinterpreted something, is this wrong? 
The problem with capturing `User_` as a whole is that `Get()` contains `VerifyPersistentStateRead`, which would (once again) require us to run explicitly in Automaton/LocalRead threads.

Since it is obviously a concern, I want to hear your thoughts on how to do all of this best.

[the remark this answers is not in the document]
Comment (raundn) [reply yt/yt/server/master/object_server/object_service.cpp:1612]: Let's break it down.

1. `securityManager->ThrottleUser` being thread-safe is fine and I don't see a better way to express in code what we're doing here.


2. The point you're making about strong-capturing `this` and, indirectly, preventing destruction of the user object if flawed, sadly. What we're asctually doing is prolonging the lifetime of the *ephemeral pointer.* That pointer, however, does not prevent the user object from being destroyed. An ephemeral pointer is, semantically, a weak one: it prolongs the lifetime of the control block but not the whole object. Thus, right before dereferencing the pointer, one has to check the refcounters to determine whether the object is still alive. Here's the sad part: those refcounters may only be read from persistent-state-read threads (i.e. either the Automaton or LocalRead).

    Basically, this is the long-winded way of saying "unless you strong-ref the user, checking its liveness must be done from persistent-state-read threads only". (And strong-reffing the user is a no-go since that would delay the destruction of a removed used indefinitely - for as long as it continues issuing requests.)

3. There's a discussion to be had about what could be expressed by the question "shouldn't we call `::Throttle` on both the global and per-user throttlers simultaneously instead of sequentially"? It appears you've had this discussion with Grigory but I just wanted to note that changing you mind on this *could* solve the issue.

File: yt/yt/server/master/security_server/request_tracker.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (raundn) [reply yt/yt/server/master/security_server/request_tracker.cpp:176]: I don't like floating point arithmetic much and would've preferred something along the lines of `std::max(1, whatever)`. But seeing as the throttler already uses doubles internally, I don't mind.

File: yt/yt/tests/integration/master/test_users.py (hunk 2 above)
Comment (raundn) [reply yt/yt/tests/integration/master/test_users.py:914]: It definitely is. 
https://github.com/ytsaurus/ytsaurus/blob/main/yt/yt/tests/library/yt_env_setup.py#L1796

File: yt/yt/server/master/object_server/object_service.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1612]: Coming back to this, I think the simplest way is to reconsider calling the throttlers sequentially vs simultaneously. There is the annoying issue of keeping the underlying promise within AllSet alive, but I guess there is no smart way around it and I'll just store the promise in the session object. I will make a fast path when both futures are already set though.

[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1612]: Ah, I remember now, the callback we subscribe with has a strong ref to TExecuteSession, so we cannot store the promise inside the session. 😞

File: yt/yt/server/master/object_server/config.h [hunk 5]
```diff
@@ -185,6 +185,15 @@ class TDynamicObjectServiceConfig
 
     bool EnableReadRequestComplexityLimits;
 
+    //! This throttler controls the rate of local write ObjectService.Execute subrequests
+    //! from *all* users except root (!) that are allowed to pass through to be scheduled
+    //! into the Automaton mutation queue.
+    //! It can be used as a form of mutation congestion control to limit the size of the
+    //! mutation queue insofar as it is induced by users' Execute requests.
+    //! Per-user request throttling is performed at the same level, but this throttler will
+    //! be acquired first.
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/config.h:194]: Changed the wording, since they are now acquired simultaneously.

File: yt/yt/server/master/object_server/object_service.cpp [hunk 6]
```diff
@@ -1600,9 +1603,20 @@ class TObjectService::TExecuteSession
                 }
 
                 const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
 
-                if (!WaitForAndContinue(result)) {
+                auto automatonThrottlerFuture = VoidFuture;
+                if (subrequestType == EExecutionSessionSubrequestType::LocalWrite && User_.Get() != securityManager->GetRootUser()) {
+                    automatonThrottlerFuture = Owner_->AutomatonRequestThrottler_->Throttle(1);
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1609]: Moved to a separate function with templates (lambdas with template arguments aren't very pretty).

File: yt/yt/server/master/object_server/object_service.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1612]: Not sure why I didn't think of this sooner, but we can get both futures and chain them with an Apply as a poor-man's replacement for AllSet. This way nobody has to own a new promise. The underlying promises of both throttler futures are referenced by the corresponding throttlers and everyone is happy this way.

File: yt/yt/tests/integration/master/test_users.py (hunk 2 above)
Comment (achulkov2) [reply yt/yt/tests/integration/master/test_users.py:914]: Removed unnecessary logic, everything seems to work fine.

File: yt/yt/server/master/object_server/object_service.cpp [hunk 7]
```diff
@@ -1600,9 +1603,20 @@ class TObjectService::TExecuteSession
                 }
 
                 const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
 
-                if (!WaitForAndContinue(result)) {
+                auto automatonThrottlerFuture = VoidFuture;
+                if (subrequestType == EExecutionSessionSubrequestType::LocalWrite && User_.Get() != securityManager->GetRootUser()) {
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/yt/server/master/object_server/object_service.cpp:1608]: Moved to a separate function with templates (lambdas with template arguments aren't very pretty).

File: yt/yt/server/master/object_server/config.h [hunk 8]
```diff
@@ -185,6 +185,14 @@ class TDynamicObjectServiceConfig
 
     bool EnableReadRequestComplexityLimits;
 
+    //! This throttler controls the rate of local write ObjectService.Execute subrequests
+    //! from *all* users except root (!) that are allowed to pass through to be scheduled
+    //! into the Automaton mutation queue.
+    //! It can be used as a form of mutation congestion control to limit the size of the
+    //! mutation queue insofar as it is induced by users' Execute requests.
+    //! This throttler is acquired simultaneously will per-user request throttling.
```
Comment (raundn) [inline yt/yt/server/master/object_server/config.h:193]: is acquired simultaneously with

File: yt/yt/server/master/object_server/object_service.cpp [hunk 9]
```diff
@@ -1582,49 +1585,53 @@ class TObjectService::TExecuteSession
...
-                const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
-                if (!WaitForAndContinue(result)) {
-                    YT_LOG_DEBUG("Throttling subrequest (User: %v, SubrequestIndex: %v, SubrequestType: %v, WorkloadType: %v)",
-                        User_->GetName(),
-                        subrequestIndex,
-                        subrequestType,
-                        workloadType);
-                    return false;
+            auto throttlerFuture = securityManager->ThrottleUser(User_.Get(), 1, WorkloadType);;
```
Comment (raundn) [inline yt/yt/server/master/object_server/object_service.cpp:1599]: extra semicolon

File: yt/yt/server/master/object_server/object_service.cpp [hunk 10]
```diff
@@ -1582,49 +1585,53 @@ class TObjectService::TExecuteSession
...
+
+            if constexpr (SubrequestType == EExecutionSessionSubrequestType::LocalWrite) {
+                if (User_.Get() != securityManager->GetRootUser()) {
+                    throttlerFuture = throttlerFuture.Apply(
+                        BIND_NO_PROPAGATE([localWriteThrottlerFuture = Owner_->LocalWriteRequestThrottler_->Throttle(1)] {
+                            return localWriteThrottlerFuture;
```
Comment (raundn) [inline yt/yt/server/master/object_server/object_service.cpp:1605]: I think this merits a commentary explaining out past difficulties with the user object lifetime here.

**Review by raundn — APPROVED** [review]

**Review by raundn — APPROVED** [review]
